File: 1601-maximum-number-of-achievable-transfer-requests/1601-maximum-number-of-achievable-transfer-requests.py
class Solution:
    ans = 0
    def maxRequest(self, requests: List[List[int]], indegree: List[int], n: int, start: int, count: int) -> None:
        if (start == len(requests)):
            for i in range(n):
                if indegree[i]:
                    return
            
            self.ans = max(self.ans, count)
            return
        
        indegree[requests[start][0]] -= 1
        indegree[requests[start][1]] += 1
        
        self.maxRequest(requests, indegree, n, start + 1, count + 1)
        
        indegree[requests[start][0]] += 1
        indegree[requests[start][1]] -= 1
        
        self.maxRequest(requests, indegree, n, start + 1, count)
    
    def maximumRequests(self, n: int, requests: List[List[int]]) -> int:
        indegree = [0] * n
        self.maxRequest(requests, indegree, n, 0, 0)
        
        return self.ans  

File: 1601-maximum-number-of-achievable-transfer-requests/1601-maximum-number-of-achievable-transfer-requests.cpp
class Solution {
public:
    int maximumRequests(int n, vector<vector<int>>& requests) {
        int ans = 0;
        for (int mask = 0; mask < (1 << requests.size()); ++mask){
            vector<int> indegree(n, 0);
            int pos = requests.size() - 1;
            int bitCount = __builtin_popcount(mask);
            
            if (bitCount <= ans)
                continue;
            
            for (int curr = mask; curr > 0 ; curr >>= 1, --pos){
                if (curr & 1){
                    --indegree[requests[pos][0]];
                    ++indegree[requests[pos][1]];
                }
            }
            
            int flag = 1;
            for (int i = 0; i < n; ++i){
                if (indegree[i]){
                    flag = 0;
                    break;
                }
            }
            
            if (flag) ans = bitCount;
        }
        
        return ans;
    }
};